compiler/riscv/codegen/FPTreeEvaluator.cpp

```diff
@@ -585,6 +585,103 @@ compareNotEqualHelper(TR::Node *node, TR::InstOpCode::Mnemonic op, TR::CodeGener
    return trgReg;
    }
 
+static TR::Register *
+compareUnorderedHelper(TR::Node *node, TR::InstOpCode::Mnemonic cmpOp, bool reverse, TR::CodeGenerator *cg)
+   {
+   TR::Node *firstChild = node->getFirstChild();
+   TR::Register *src1Reg = cg->evaluate(firstChild);
+   TR::Node *secondChild = node->getSecondChild();
+   TR::Register *src2Reg = cg->evaluate(secondChild);
+
+   TR::Register *trgReg = cg->allocateRegister();
+
+   /*
+    * To avoid explicit NaN checks, we perform opposite comparison
+    * and flip the result, i.e, when computing "(x >= y) or unordered",
+    * actually compute "not (x < y)"
+    *
+    * The cmpOp passed as parameter is already the opposite comparison.
+    */
+
+   if (reverse)
+      generateRTYPE(cmpOp, node, trgReg, src2Reg, src1Reg, cg);
+   else
+      generateRTYPE(cmpOp, node, trgReg, src1Reg, src2Reg, cg);
+
+   generateITYPE(TR::InstOpCode::_xori, node, trgReg, trgReg, 1, cg);
+
+   cg->decReferenceCount(firstChild);
+   cg->decReferenceCount(secondChild);
+   node->setRegister(trgReg);
+   return trgReg;
+   }
+
+static TR::Register *
+compareEqualOrUnorderedHelper(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   TR::Node *firstChild = node->getFirstChild();
+   TR::Register *src1Reg = cg->evaluate(firstChild);
+   TR::Node *secondChild = node->getSecondChild();
+   TR::Register *src2Reg = cg->evaluate(secondChild);
+
+   TR::Register *trgReg = cg->allocateRegister();
+
+   /*
+    * Here we exploit the fact that exactly one of the following conditions
+    * must always be true:
+    *
+    *   1. x < y
+    *   2. x > y
+    *   3. x == y
+    *   4  x and y are unordered
+    *
+    *   We want to check whether conditions 3 or 4 are true, but we can as well check that
+    *   neither 1 nor 2 are true and flip the result, i.e., compute !(x < y || x > y).
```

Comment:
There seems to be a problematic double negation going on in this sentence. Checking that neither 1 nor 2 are true would imply computing `!(x < y || x > y)` and flipping that result would seem to imply that this code should be computing `x < y || x > y`.

Reply:
True, will fix.